File: interflow-frontend/src/model/stock.ts
import { Matiere, MatiereModel } from './matiere';

export interface Stock {
  id?: string;
  article: string; // Clé primaire
  libelle_article: string;
  du: string;
  quantite: number;
  udm: string;
  statut_lot: string;
  division: string;
  magasin: string; // Clé primaire
  emplacement: string; // Clé primaire
  contenant: string; // Clé primaire
  statut_proprete: string;
  reutilisable: string;
  statut_contenant: string;
  classification?: string | null;
  restriction?: string | null;
  lot_fournisseur?: string | null;
  capacite?: string | null;
  commentaire?: string | null;
  date_creation?: Date | null;
  dluo?: Date | null;
  matiere?: Matiere | null;
}

export class StockModel implements Stock {
  id?: string;
  article: string;
  libelle_article: string;
  du: string;
  quantite: number;
  udm: string;
  statut_lot: string;
  division: string;
  magasin: string;
  emplacement: string;
  contenant: string;
  statut_proprete: string;
  reutilisable: string;
  statut_contenant: string;
  classification?: string | null;
  restriction?: string | null;
  lot_fournisseur?: string | null;
  capacite?: string | null;
  commentaire?: string | null;
  date_creation?: Date | null;
  dluo?: Date | null;
  matiere?: Matiere | null;

  constructor(data: Partial<Stock>) {
    this.article = data.article || "";
    this.libelle_article = data.libelle_article || "";
    this.du = data.du || "";
    this.quantite = data.quantite || 0;
    this.udm = data.udm || "";
    this.statut_lot = data.statut_lot || "";
    this.division = data.division || "";
    this.magasin = data.magasin || "";
    this.emplacement = data.emplacement || "";
    this.contenant = data.contenant || "";
    this.statut_proprete = data.statut_proprete || "";
    this.reutilisable = data.reutilisable || "";
    this.statut_contenant = data.statut_contenant || "";
    this.classification = data.classification || null;
    this.restriction = data.restriction || null;
    this.lot_fournisseur = data.lot_fournisseur || null;
    this.capacite = data.capacite || null;
    this.commentaire = data.commentaire || null;
    this.date_creation = data.date_creation || null;
    this.dluo = data.dluo || null;
    
    // La matière sera initialisée si nécessaire
    if (!data.matiere) {
      // Créer une matière temporaire avec le libellé
      this.matiere = new MatiereModel({ code_mp: "TOBEDEFINED", nom: this.libelle_article });
    } else {
      this.matiere = data.matiere;
    }
    
    // Générer une clé primaire composite basée sur le tuple unique
    this.id = `${this.article}_${this.magasin}_${this.emplacement}_${this.contenant}`;
  }

  static fromData(data: Record<string, unknown>): StockModel {
    const processedData = { ...data };
    
    // Gestion des dates
    if (processedData.date_creation && typeof processedData.date_creation === 'string') {
      try {
        processedData.date_creation = new Date(processedData.date_creation);
      } catch {
        processedData.date_creation = null;
      }
    }
    
    if (processedData.dluo && typeof processedData.dluo === 'string') {
      try {
        processedData.dluo = new Date(processedData.dluo);
      } catch {
        processedData.dluo = null;
      }
    }
    
    // Gestion de la matière
    if (processedData.matiere && typeof processedData.matiere === 'object') {
      processedData.matiere = MatiereModel.fromData(processedData.matiere as Record<string, unknown>);
    }
    
    return new StockModel(processedData);
  }

  toData(): Record<string, unknown> {
    return {
      id: this.id,
      article: this.article,
      libelle_article: this.libelle_article,
      du: this.du,
      quantite: this.quantite,
      udm: this.udm,
      statut_lot: this.statut_lot,
      division: this.division,
      magasin: this.magasin,
      emplacement: this.emplacement,
      contenant: this.contenant,
      statut_proprete: this.statut_proprete,
      reutilisable: this.reutilisable,
      statut_contenant: this.statut_contenant,
      classification: this.classification,
      restriction: this.restriction,
      lot_fournisseur: this.lot_fournisseur,
      capacite: this.capacite,
      commentaire: this.commentaire,
      date_creation: this.date_creation?.toISOString() || null,
      dluo: this.dluo?.toISOString() || null,
      matiere: this.matiere instanceof MatiereModel ? this.matiere.toData() : this.matiere
    };
  }
} 

File: interflow-backend/src/models/stock.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List, Iterator, Union
from datetime import datetime
from models.matieres import Matiere

class Stock(BaseModel):
    id: Optional[str] = Field(None, description="Identifiant unique du stock (généré automatiquement)")
    article: str = Field(..., description="Code de l'article")
    libelle_article: str = Field(..., description="Libellé de l'article")
    du: Optional[str] = Field(None, description="Date d'utilisation")
    quantite: float = Field(..., description="Quantité disponible", ge=0)
    udm: str = Field(..., description="Unité de mesure")
    statut_lot: str = Field(..., description="Statut du lot")
    division: str = Field(..., description="Division")
    magasin: str = Field(..., description="Code du magasin")
    emplacement: str = Field(..., description="Emplacement dans le magasin")
    contenant: str = Field(..., description="Contenant")
    statut_proprete: str = Field(..., description="Statut de propreté")
    reutilisable: str = Field(..., description="Réutilisable ou non")
    statut_contenant: str = Field(..., description="Statut du contenant")
    classification: Optional[str] = Field(None, description="Classification")
    restriction: Optional[str] = Field(None, description="Restrictions")
    lot_fournisseur: Optional[str] = Field(None, description="Lot fournisseur")
    capacite: Optional[float] = Field(None, description="Capacité")
    commentaire: Optional[str] = Field(None, description="Commentaire")
    date_creation: Optional[datetime] = Field(None, description="Date de création")
    dluo: Optional[datetime] = Field(None, description="Date limite d'utilisation optimale")
    matiere: Matiere | None = Field(None, description="Matière associée")

    @field_validator('division', mode='before')
    @classmethod
    def normalize_division(cls, v):
        """Normalise les codes division (supprime les zéros de début)"""
        if not v:
            return ""
        value_str = str(v).strip()
        if value_str.isdigit() and len(value_str) > 1:
            normalized = str(int(value_str))
            if len(normalized) <= 4:  # Codes division max 4 chiffres
                return normalized
        return value_str

    @field_validator('lot_fournisseur', mode='before')
    @classmethod
    def normalize_lot_fournisseur(cls, v):
        """Normalise les lots fournisseur (supprime .0 des entiers)"""
        if not v:
            return None
        value_str = str(v).strip()
        # Supprimer .0 si c'est un entier
        if value_str.endswith('.0') and value_str.replace('.0', '').isdigit():
            return value_str.replace('.0', '')
        return value_str

    @field_validator('capacite', 'quantite', mode='before')
    @classmethod
    def normalize_numeric_values(cls, v):
        """Normalise les valeurs numériques (convertit les virgules en points)"""
        if v is None:
            return None if 'capacite' in cls.__name__ else 0.0

        # Si c'est déjà un nombre, le garder
        if isinstance(v, (int, float)):
            return float(v)

        value_str = str(v).strip()

        # Convertir les virgules en points
        if ',' in value_str:
            # Gérer les cas spéciaux comme "500," -> "500.0"
            if value_str.endswith(','):
                value_str = value_str[:-1] + '.0'
            else:
                value_str = value_str.replace(',', '.')

        try:
            return float(value_str)
        except ValueError:
            return None if 'capacite' in cls.__name__ else 0.0

    @field_validator('restriction', 'classification', 'commentaire', 'du', mode='before')
    @classmethod
    def clean_string_fields(cls, v):
        """Nettoie les champs string pour gérer les valeurs NaN"""
        if v is None:
            return None

        # Si c'est déjà une string valide, la garder
        if isinstance(v, str):
            value_str = v.strip()
            if value_str.lower() in ['nan', 'none', 'null', '']:
                return None
            return value_str

        # Si c'est un float NaN, retourner None
        if isinstance(v, float) and str(v).lower() == 'nan':
            return None

        # Convertir en string et nettoyer
        value_str = str(v).strip()
        if value_str.lower() in ['nan', 'none', 'null', '']:
            return None

        return value_str

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        # La matière sera initialisée par le repository si nécessaire
        if not self.matiere:
            # Créer une matière temporaire avec le libellé
            self.matiere = Matiere(code_mp="TOBEDEFINED", nom=self.libelle_article)
        # Générer une clé primaire composite basée sur le tuple unique
        self.id = f"{self.article}_{self.magasin}_{self.emplacement}_{self.contenant}"

    @classmethod
    def from_model_dump(cls, data: dict):
        return cls(**data)
